### PR TITLE
Fixed PowerAuthPassword for non-ascii chars

### PR DESCRIPTION
**Pull request by Hopsaheysa**
#310

### CODE REVIEW

File: packages/react-native-powerauth-mobile-sdk/android/src/main/java/com/wultra/android/powerauth/js/PasswordCodePointScheme.kt [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    // this library's minSdkVersion 21.
+    val result = ArrayList<Int>(normalized.length)
+    var i = 0
+    while (i < normalized.length) {
+        val codePoint = normalized.codePointAt(i)
+        result.add(codePoint)
+        i += Character.charCount(codePoint)
+    }
+    return result
+}
```
[the remark this answers is not in the document]
Comment (Hopsaheysa) [reply packages/react-native-powerauth-mobile-sdk/android/src/main/java/com/wultra/android/powerauth/js/PasswordCodePointScheme.kt:95]: rejected sooner in withPassword

File: packages/react-native-powerauth-mobile-sdk/android/src/main/java/com/wultra/android/powerauth/js/PowerAuthPasswordJsModule.kt [hunk 2]
```diff
@@ -245,40 +265,59 @@ class PowerAuthPasswordJsModule(private val objectRegister: ObjectRegisterJs) :
...
-        action: CharacterAction
+        action: CodePointsAction
     ) {
-        if (character < 0 || character > Constants.CODEPOINT_MAX) {
-            promise.reject(Errors.EC_WRONG_PARAMETER, "Invalid CodePoint")
+        if (codePoints.size() == 0) {
+            promise.reject(Errors.EC_WRONG_PARAMETER, "Empty code points array")
             return
         }
+        val points = ArrayList<Int>(codePoints.size())
```
Comment (smanikw) [inline packages/react-native-powerauth-mobile-sdk/android/src/main/java/com/wultra/android/powerauth/js/PowerAuthPasswordJsModule.kt:305]: why is this withPassword overload doing so much?

File: packages/react-native-powerauth-mobile-sdk/android/src/main/java/com/wultra/android/powerauth/js/PowerAuthJsModule.kt [hunk 3]
```diff
@@ -366,6 +367,7 @@ class PowerAuthJsModule(
                 val auth: PowerAuthAuthentication = constructAuthentication(authMap, true, false)
                 val result: Int = sdk.persistActivationWithAuthentication(context, auth)
                 if (result == PowerAuthErrorCodes.SUCCEED) {
+                    markActivationWithCorrectedPasswordScheme(context, instanceId, objectRegister)
```
Comment (smanikw) [inline packages/react-native-powerauth-mobile-sdk/android/src/main/java/com/wultra/android/powerauth/js/PowerAuthJsModule.kt:370]: there is an edge case, when the activation was persisted with a password created with the constructor, rather than the createPassword() method. The docs still support the constructor way.

this test shows it:
```typescript
 async testOwnerlessPasswordSchemeSurvivesActivation() {
        const sdk = this.sdk
        expect(await sdk.hasValidActivation()).toBe(false)

        const originalPassword = new PowerAuthPassword()
        await originalPassword.addCharacter(this.familyEmoji)
        expect(await originalPassword.length()).toBe(1) // Legacy scheme keeps only U+1F468 (man).

        const detail = await this.helper.createActivation()
        await sdk.createActivation(PowerAuthActivation.createWithActivationCode(detail.activationCode, 'RN'))
        await sdk.persistActivation(PowerAuthAuthentication.persistWithPassword(originalPassword))
        await this.helper.commitActivation()
        expect(await sdk.hasValidActivation()).toBe(true)

        const passwordCheck = sdk.createPassword()
        await passwordCheck.addCharacter(this.familyEmoji)
        expect(await passwordCheck.length()).toBe(5) // Corrected scheme keeps the complete ZWJ sequence.

        // This currently fails with AUTHENTICATION_ERROR because the activation was enrolled with the
        // legacy one-code-point representation but was unconditionally marked as corrected.
        await sdk.validatePassword(passwordCheck)
    }
```

File: packages/react-native-powerauth-mobile-sdk/android/src/main/java/com/wultra/android/powerauth/js/PasswordCodePointScheme.kt [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    val activationId = sdk.activationIdentifier ?: return false
+    return prefs(context).getBoolean(KEY_PREFIX + activationId, false)
+}
+
+/**
+ * Returns the NFC-normalized form of the given raw Unicode code points (e.g. composing a base letter +
+ * combining mark into one code point where possible; sequences with no NFC rule, like ZWJ emoji, are
+ * unchanged). Only used for the corrected scheme - legacy always stores the raw 1st. code point as-is.
+ */
+fun nfcNormalizeCodePoints(codePoints: List<Int>): List<Int> {
```
Comment (smanikw) [inline packages/react-native-powerauth-mobile-sdk/android/src/main/java/com/wultra/android/powerauth/js/PasswordCodePointScheme.kt:82]: what about this case?
addCharacter("e\u0301")

addCharacter("e")
addCharacter("\u0301")

this would produce two different passwords, are we good with this? i worry this would create unexpected problems.

we can see that in fromString() function.

example test:
```typescript
async testFromStringAppliesTheSameNfcNormalization() {
        const config = new PowerAuthConfiguration('ARAVst+fkgOOT/U1gBr1qLMDEOTfEduuLUvbpOmTq7cI+skBAUEEVjKe+8yFg62GvhwU8eE3iEZZCOeNqtEyz2AXXs/yZewnmdETC8J2sNcw5NnIApYDUmBh2n+XRHize4EiVdetjQ==', 'http://localhost/wrong')
        const powerAuth = new PowerAuth(this.getRandomId())
        this.cleanup.push(powerAuth)
        await powerAuth.configure(config)
        expect(await powerAuth.hasValidActivation()).toBe(false)

        const decomposedEAcute = "\u0065\u0301"
        const createdFromString = await PowerAuthPassword.fromString(
            decomposedEAcute,
            true,
            undefined,
            powerAuth.instanceId
        )
        const addedInOneCall = powerAuth.createPassword()
        this.cleanup.push(createdFromString, addedInOneCall)

        await addedInOneCall.addCharacter(decomposedEAcute)

        expect(await createdFromString.length()).toBe(2) // U+0065 followed by U+0301
        expect(await addedInOneCall.length()).toBe(1)     // NFC-composed U+00E9

        // Equivalent construction APIs should produce identical password bytes. This currently fails
        // because fromString() normalizes each code point separately rather than the complete string.
        expect(await createdFromString.isEqualTo(addedInOneCall)).toBe(true)
    }
```

both cases have their disadvantages so we should brainstorm about this

File: packages/react-native-powerauth-mobile-sdk/ios/PowerAuth/PasswordCodePointScheme.m [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+static PowerAuthSDK * PACPS_FindSdk(NSString * instanceId, PowerAuthObjectRegister * objectRegister)
+{
+    if (!instanceId) {
+        return nil;
+    }
+    return [objectRegister findObjectWithId:instanceId expectedClass:[PowerAuthSDK class]];
+}
+
+void PACPS_MarkActivationWithCorrectedPasswordScheme(NSString * instanceId, PowerAuthObjectRegister * objectRegister)
```
Comment (smanikw) [inline packages/react-native-powerauth-mobile-sdk/ios/PowerAuth/PasswordCodePointScheme.m:35]: what happens when activation sharing ?

**Review by smanikw — CHANGES_REQUESTED** [review]
:wrench: :wrench: :wrench:

Image

File: packages/react-native-powerauth-mobile-sdk/android/src/main/java/com/wultra/android/powerauth/js/PasswordCodePointScheme.kt (hunk 4 above)
Comment (Hopsaheysa) [reply packages/react-native-powerauth-mobile-sdk/android/src/main/java/com/wultra/android/powerauth/js/PasswordCodePointScheme.kt:82]: true... idk what more than the current docs I could do:

> Note that this is not recommended. Do this only when you retrieve the whole string from a text input.
